modules/dataquery/php/endpoints/queries/query/run.class.inc

```diff
@@ -55,48 +60,49 @@ class Run extends \LORIS\Http\Endpoint
      */
     public function runQuery(\User $user, int $queryID) : ResponseInterface
     {
-        try {
-            $query = new \LORIS\dataquery\Query($this->loris, $queryID);
+        $query = new \LORIS\dataquery\Query($this->loris, $queryID);
 
-            $queryRun = $query->newRun($user);
-            $queryRun->insertCandidates($user);
-            $data = $queryRun->getQueryDataProvisioner();
+        $db = $this->loris->getDatabaseConnection();
+        // Put the query run in a transaction to make it atomic
+        $queryRun = $query->newRun($user);
+        $db->beginTransaction();
+        $data = $queryRun->getQueryDataProvisioner()->execute($user);
 
-            $table  = (new \LORIS\Data\Table())
-                ->withDataFrom($data);
-            $rows   = $table->getRows($user);
-            $stream = new \LORIS\Http\DataIteratorBinaryStream($rows);
-            return (new \LORIS\Http\Response())
-                ->withHeader("Content-Type", "text/plain")
-                //->withHeader("Content-Type", "application/octet-stream")
-                ->withBody($stream);
-        } catch (\LorisException $e) {
-            return new \LORIS\Http\Response\JSON\NotImplemented($e->getMessage());
+        $values = [];
 
-        }
-    }
+        set_time_limit(300);
+        $runid  = $queryRun->getRunId();
+        $stream = new \LORIS\Http\DataIteratorBinaryStream(
+            $data,
+            function (
+                $candid,
+                $rowsArray,
+                /* @phan-suppress-next-line PhanUnusedClosureParameter */
+                $rowsBinary,
+                bool $end
+            ) use (&$values, &$db, &$runid) {
+                $values[] = join(
+                    ',',
+                    [$runid, $candid, $db->quote(json_encode($rowsArray))]
+                );
+                if ((count($values) >= 2000 || $end) && count($values) > 0) {
```

Comment:
Also wondering if this value should be configurable - why 2000?

Reply:
It is arbitrary but I don't think it needs to be configurable.. the goal is to break it up into chunks so that the insert is done in batches but the SQL query string doesn't grow without bound for large queries. 2000 comes from trial and error.